cpp/include/kvikio/shim/libcurl.hpp

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#ifndef KVIKIO_LIBCURL_FOUND
+#error \
+  "cannot include the remote IO API, please build KvikIO with libcurl (-DKvikIO_REMOTE_SUPPORT=ON)"
+#endif
+
+#include <cstring>
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include <curl/curl.h>
+
+#include <kvikio/defaults.hpp>
+#include <kvikio/error.hpp>
+#include <kvikio/parallel_operation.hpp>
+#include <kvikio/posix_io.hpp>
+#include <kvikio/utils.hpp>
+
+namespace kvikio {
+
+/**
+ * @brief Singleton class to initialize and cleanup the global state of libcurl
+ *
+ * Notice, libcurl allows the use of a singleton class:
+ *
+ * In a C++ module, it is common to deal with the global constant situation by defining a special
+ * class that represents the global constant environment of the module. A program always has exactly
+ * one object of the class, in static storage. That way, the program automatically calls the
+ * constructor of the object as the program starts up and the destructor as it terminates. As the
+ * author of this libcurl-using module, you can make the constructor call curl_global_init and the
+ * destructor call curl_global_cleanup and satisfy libcurl's requirements without your user having
+ * to think about it. (Caveat: If you are initializing libcurl from a Windows DLL you should not
+ * initialize it from DllMain or a static initializer because Windows holds the loader lock during
+ * that time and it could cause a deadlock.)
+ *
+ * Source <https://curl.se/libcurl/c/libcurl.html>.
+ */
+class LibCurl {
+ public:
+  // We hold an unique pointer to the raw curl handle and sets `curl_easy_cleanup` as its Deleter.
+  using UniqueHandlePtr = std::unique_ptr<CURL, std::function<void(CURL*)>>;
+
+ private:
+  std::mutex _mutex{};
+  // Curl handles free to be used.
+  std::vector<UniqueHandlePtr> _free_curl_handles{};
```

Comment:
*question*: In theory, the caller could create endless curl handles and never release them, do we care about that case?

Reply:
No :)